src/data_readers/unit_test/data_reader_synthetic_test_public_api.cpp

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2014-2021, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include <catch2/catch.hpp>
+
+#include "MPITestHelpers.hpp"
+#include "TestHelpers.hpp"
+#include "lbann/proto/proto_common.hpp"
+#include <google/protobuf/text_format.h>
+#include <lbann.pb.h>
+
+#include <cstdlib>
+#include <errno.h>
+#include <string.h>
+
+//#include "./data_reader_common_catch2.hpp"
+#include "lbann/data_readers/data_reader_synthetic.hpp"
+#include "lbann/data_readers/utils/input_data_type.hpp"
+#include "lbann/utils/threads/thread_pool.hpp"
+#include "lbann/utils/threads/thread_utils.hpp"
+#include "lbann/utils/hash.hpp"
+
+class DataReaderSyntheticWhiteboxTester
+{
+public:
+  bool fetch_datum(lbann::data_reader_synthetic& dr, lbann::CPUMat& X, int data_id, int mb_idx) {
+    return dr.fetch_datum(X, data_id, mb_idx);
+  }
+  bool fetch_label(lbann::data_reader_synthetic& dr, lbann::CPUMat& Y, int data_id, int mb_idx) {
+    return dr.fetch_label(Y, data_id, mb_idx);
+  }
+  bool fetch_response(lbann::data_reader_synthetic& dr, lbann::CPUMat& Y, int data_id, int mb_idx) {
+    return dr.fetch_response(Y, data_id, mb_idx);
+  }
+};
+
+TEST_CASE("Synthetic data reader public API tests",
+          "[mpi][data_reader][synthetic][public]")
+{
+  // initialize stuff (boilerplate)
+  auto& comm = unit_test::utilities::current_world_comm();
+  lbann::init_random(42, 1);
+  lbann::init_data_seq_random(42);
+
+  // Create a local copy of the RNG to check the synthetic data reader
+  lbann::fast_rng_gen ref_fast_generator;
+  ref_fast_generator.seed(lbann::hash_combine(42, 0));
+
+  // Initalize a per-trainer I/O thread pool
+  auto io_thread_pool = lbann::make_unique<lbann::thread_pool>();
+  io_thread_pool->launch_pinned_threads(1, 1);
+
+  std::set<std::string> active_data_fields = {"samples"};
+  active_data_fields.insert(GENERATE("labels", "responses"));
```

Comment:
```suggestion
  active_data_fields.insert(GENERATE(std::string("labels"),
                                     std::string("responses")));
```
Apple clang was whining about this one...